attestation-container/attest/attestation.go

```diff
@@ -0,0 +1,150 @@
+package attest
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+// Data structures are based on SEV-SNP Firmware ABI Specification
+// https://www.amd.com/en/support/tech-docs/sev-secure-nested-paging-firmware-abi-specification
+
+const (
+	ATTESTATION_REPORT_SIZE = 1184 // Size of ATTESTATION_REPORT (Table 21)
+	REPORT_DATA_SIZE        = 64   // Size of REPORT_DATA_SIZE in ATTESTATION_REPORT
+	REPORT_REQ_SIZE         = 96   // Size of MSG_REPORT_REQ (Table 20)
+	REPORT_RSP_SIZE         = 1280 // Size of MSG_REPORT_RSP (Table 23)
+	PAYLOAD_SIZE            = 40   // Size of sev_snp_guest_request struct from sev-snp driver include/uapi/linux/psp-sev-guest.h
+)
+
+// Message Type Encodings (Table 100)
+const (
+	MSG_REPORT_REQ = 5
+	MSG_REPORT_RSP = 6
+)
+
+// From sev-snp driver include/uapi/linux/psp-sev-guest.h
+const SEV_SNP_GUEST_MSG_REPORT = 3223868161
+
+const SNP_DEVICE_PATH = "/dev/sev"
+
+/*
+Creates and returns MSG_REPORT_REQ message bytes (SEV-SNP Firmware ABI Specification Table 20)
+*/
+func createReportReqBytes(reportData [REPORT_DATA_SIZE]byte) [REPORT_REQ_SIZE]byte {
+	reportReqBytes := [REPORT_REQ_SIZE]byte{}
+	for i := 0; i < REPORT_DATA_SIZE; i++ {
+		reportReqBytes[i] = reportData[i]
+	}
+	return reportReqBytes
+}
+
+/*
+Creates and returns byte array of the following C struct
+
+// From sev-snp driver include/uapi/linux/psp-sev-guest.h
+// struct sev_snp_guest_request {
+//   uint8_t req_msg_type;
+//   uint8_t rsp_msg_type;
+//   uint8_t msg_version;
+//   uint16_t request_len;
+//   uint64_t request_uaddr;
+//   uint16_t response_len;
+//   uint64_t response_uaddr;
+//   uint32_t error;		// firmware error code on failure (see psp-sev.h)
+// };
+
+The padding is based on Section 3.1.2 of System V ABI for AMD64
+https://www.uclibc.org/docs/psABI-x86_64.pdf
+*/
+func createPayloadBytes(reportReqPtr uintptr, ReportRespPtr uintptr) ([PAYLOAD_SIZE]byte, error) {
+	payload := [PAYLOAD_SIZE]byte{}
+	var buf bytes.Buffer
+	// req_msg_type
+	if err := binary.Write(&buf, binary.LittleEndian, uint8(MSG_REPORT_REQ)); err != nil {
+		return payload, err
```

Comment:
Why do we return `payload, err` in these branches, rather than `nil, err`? I don't know which is more idiomatic Go, but other instances in this PR use `nil, err`, and it feels like that will avoids bugs where callers check the wrong field.

Reply:
I didn't consider it well. I agree to make it `nil, err`

Reply:
Ah it's impossible to return `nil`, since its return type is array not slice.